subsys/shell/shell.c

```diff
@@ -1668,15 +1668,25 @@ static void help_options_print(const struct shell *shell,
 
 	/* Formating and printing all available options (except -h, --help). */
 	for (size_t i = 0; i < opt_cnt; ++i) {
+		size_t len = 0;
+		size_t buf_len = sizeof(shell->ctx->temp_buff);
 		if (opt[i].optname_short) {
-			strcpy(shell->ctx->temp_buff, opt[i].optname_short);
+			len += shell_strlen(opt[i].optname_short);
```

Comment:
This is very good catch!

I am not sure if this code solves the problem. In my opinion if command buffer is not large enough to store option or command user will not have chance to use it. So it makes limited sense to print only partial option string with no indication it is not compleate.

I think we can implement either an ASSERT or shell error message that command buffer is too small. What do you think?

Reply:
yeah, I was not sure about the behaviour here as well, As it is just to print a help, I thought would be better some information than nothing, but we can assert or return an error. What do you prefer ?

Reply:
Below is my proposal, if buffer is too small shell will print warning and as many data as possible:

```
static void help_options_print(const struct shell *shell,
			       const struct shell_getopt_option *opt,
			       size_t opt_cnt)
{
	u32_t buf_len = sizeof(shell->ctx->temp_buff);
	static const char help_opt[] = "-h, --help";
	u32_t longest_name = shell_strlen(help_opt);
	static const char opt_sep[] = ", ";


	shell_fprintf(shell, SHELL_NORMAL, "Options:\r\n");

	if ((opt == NULL) || (opt_cnt == 0)) {
		help_item_print(shell, help_opt, longest_name,
				"Show command help.");
		return;
	}

	/* Looking for the longest option string. */
	for (size_t i = 0; i < opt_cnt; ++i) {
		u32_t len = shell_strlen(opt[i].optname_short) +
						 shell_strlen(opt[i].optname) +
						 shell_strlen(opt_sep);

		longest_name = len > longest_name ? len : longest_name;
	}

	if (longest_name > buf_len) {
		shell_fprintf(shell, SHELL_WARNING,
			      "Some options will not be printed correctly."
			      "Increase command buffer size.\r\n");
	}

	/* help option is printed first */
	help_item_print(shell, help_opt, longest_name, "Show command help.");

	/* prepare a buffer to compose a string:
	 *	option name short - option separator - option name
	 */
	memset(shell->ctx->temp_buff, 0, longest_name + 1);

	/* Formating and printing all available options (except -h, --help). */
	for (size_t i = 0; i < opt_cnt; ++i) {
		u32_t len = 0;
		if (opt[i].optname_short) {
			len += shell_strlen(opt[i].optname_short);
			strncpy(shell->ctx->temp_buff, opt[i].optname_short,
				min(len, buf_len));
		}
		if (opt[i].optname && (len < buf_len)) {
			if (*shell->ctx->temp_buff) {
				len += shell_strlen(opt_sep) +
					shell_strlen(opt[i].optname);
				if (len < buf_len) {
					strcat(shell->ctx->temp_buff, opt_sep);
					strcat(shell->ctx->temp_buff,
					       opt[i].optname);
				}
			} else {
				strncpy(shell->ctx->temp_buff, opt[i].optname,
					buf_len);
			}
		}
		help_item_print(shell, shell->ctx->temp_buff, longest_name,
				opt[i].optname_help);
	}
}
```